woodpile/src/pile_reader.rs

```diff
@@ -0,0 +1,1617 @@
+        }
+    }
+}
+
+impl std::ops::DerefMut for PileReaderState {
+    #[inline(always)]
+    fn deref_mut(&mut self) -> &mut PileReader {
+        match self {
+            PileReaderState::Peek(reader) => reader,
+            PileReaderState::Pending(reader) => reader,
+            PileReaderState::Stable(reader) => reader,
+            PileReaderState::StableRecovery(reader) => reader,
+        }
+    }
+}
+
+impl std::ops::Deref for StablePileReader {
+    type Target = PileReader;
+
+    #[inline(always)]
+    fn deref(&self) -> &PileReader {
+        &self.inner
+    }
+}
+
+impl std::ops::DerefMut for StablePileReader {
+    #[inline(always)]
+    fn deref_mut(&mut self) -> &mut PileReader {
+        &mut self.inner
+    }
+}
+
+const LOG_SUFFIX: &[u8] = b".log";
+const _: () = assert!(
+    crate::LOG_EXTENSION.len() == 3
+        && crate::LOG_EXTENSION.as_bytes()[0] == b'l'
+        && crate::LOG_EXTENSION.as_bytes()[1] == b'o'
+        && crate::LOG_EXTENSION.as_bytes()[2] == b'g'
+);
+
+impl PileReader {
+    /// Creates a new `PileReader` from a vector of pairs, where
+    /// the first element is the path to the log file, and the second
+    /// is the byte offset at which we should start reading records.
+    ///
+    /// If `max_record_size` is provided, records with a decoded size
+    /// larger than that maximum value are dropped on the fly.  If
+    /// `max_record_size` is `None`, we use [`DEFAULT_MAX_RECORD_SIZE`].
+    pub fn new_from_source_files(
+        pile_dir: PathBuf,
+        source_files: Vec<(PathBuf, u64)>,
+        max_record_size: Option<usize>,
+    ) -> Self {
+        PileReader {
+            current_reader: None,
+            pile_dir,
+            next_source_file: 0,
+            source_files,
+            max_record_size: max_record_size.unwrap_or(DEFAULT_MAX_RECORD_SIZE),
+            virtual_dir: None,
+        }
+    }
+
+    /// Creates a new `PileReader` for the log files in the zip `archive`.
+    ///
+    /// On success, returns the `PileReader`, and true if any of the archived
+    /// files are shorter than expected from `start_offsets`.
+    pub fn new_from_zip_archive(
+        pile_dir: PathBuf,
+        mut archive: ZipArchive<File>,
+        start_offsets: Option<&HashMap<PathBuf, u64>>,
+        max_record_size: Option<usize>,
+    ) -> Result<(Self, bool)> {
+        use std::ffi::OsStr;
+        use std::path::Path;
+
+        let empty = HashMap::new();
+        let start_offsets = start_offsets.unwrap_or(&empty);
+
+        let mut offsets: Vec<(u64, u64)> = Vec::with_capacity(archive.len());
+        let mut source_files: Vec<(PathBuf, u64)> = Vec::with_capacity(archive.len());
+        let mut rewound = false;
+
+        // Count how many offsets we found in `start_offsets`.
+        let mut found = 0usize;
+
+        let subdir = Path::new(pile_dir.file_name().unwrap_or(OsStr::new("")));
+        for idx in 0..archive.len() {
+            let entry = archive.by_index(idx)?;
```

Comment:
Getting `data_start()` requires seeking to the local file header, which Is apparently what `by_index()` does. That's not great when coupled with huge requests sizes in NFS. We can do something smarter since we know exactly what the producer wrote down, given the file name.